### ISSUE TITLE
`numexpr` still uses `NPY_ARRAY_UPDATEIFCOPY`

### ISSUE THREAD

**Issue by seberg** [pydata/numexpr#397]
We deprecated `NPY_ARRAY_UPDATEIFCOPY ` a pretty long time ago, but with a Python warning.  There is one (probably almost never used) branch that still has an `NPY_ARRAY_UPDATEIFCOPY `.  NumPy main currently removes the symbol entirely, so this will not work.  And if compiled on older NumPy versions, it would just do the wrong thing (not copy back).

It is in an `n_inputs == 0` branch, is that branch meaningless enough, or should we fix this here, and delay the deprecation for one release in NumPy for the sake of `numexpr`?

Here is the code:
https://github.com/pydata/numexpr/blob/d745933d1fb79f08ca152463480d506199ea24f0/numexpr/interpreter.cpp#L1271

**Comment by robbmcleod** [comment]
I've managed to figure out how to replicate this case:

    import numpy as np
    import numexpr as ne
    x = np.empty(1, dtype=int)
    ne.evaluate('1', out=x)

Just `ne.evaluate('1')` doesn't go into this branch. My concern (as always) is whether it could effect Pandas or PyTables as they are downstream. 

I'm not really sure what you mean about a Python warning as I don't get one when running the above code snippet? Maybe because it's in the C-API?

If I just remove the `NPY_ARRAY_UPDATEIFCOPY` flag it seems to compile and run just fine.

**Comment by seberg** [comment]
It should give a warning, try running with `python -Wall` or something. It was a DeprecationWarning, so it may be hidden from you as an "end-user".

If you are unsure/worried, I guess we should probably revert this, and add a deprecation macro for at least one version to flush out untested code paths.

**Comment by seberg** [comment]
> If I just remove the NPY_ARRAY_UPDATEIFCOPY flag it seems to compile and run just fine.

The result should be incorrect though?  I.e. return `out` unmodified?  This has the potential of silent error, if it was just an exception, I probably would not bother.

**Comment by robbmcleod** [comment]
No, it definitely works as intended, or as I assume it was to be, since I didn't write at of the code we're talking about:

    x = np.zeros(1, dtype=int)
    print(f"pre-eval, x = {x} at {x.__array_interface__['data'][0]}")
    ne.evaluate('3', out=x)
    print(f"post-eval, x = {x} at {x.__array_interface__['data'][0]}")

outputs:

    pre-eval, x = [0] at 1866944975104
    n_inputs = 0
    operands[0] = 000001B2AF069930, retsig = i
    Case with NPY_ARRAY_UPDATEIFCOPY
    post-eval, x = [3] at 1866944975104

In which version of NumPy will this flag be removed, and when will the release be?

**Comment by seberg** [comment]
As of now, it would be the next version, i.e. 1.23.0 (or maybe 2.0 if that happens), in about 6 months.

Sorry for the noise.  Looking closer, the bug is just that the flag is used at all :). The flag only makes sense if you also set the base object (or better call `PyArray_SetUpdateIfCopyBase`), which you do not do.

So nothing wrong will happy, because the flag should just be ignored I think.

**Comment by robbmcleod** [comment]
There's no issue with `NPY_ITER_UPDATEIFCOPY`? I think not from the NumPy issues tracker, but wanted to ask to be sure.

Closing otherwise.

**Comment by seberg** [comment]
Wait, no, arrrggg.  NumPy does this for you, so I think it should be used, but here, probably only in the crazy unlikely case that the input is an unaligned array.

**Comment by seberg** [comment]
`NPY_ITER_UPDATEIFCOPY` is fine, it uses the newer `WRITEBACK_IF_COPY` internally for the arrays.  A bit confusing, but for the iterator it was possible to keep things as they were, for UPDATEIFCOPY it was not.

**Comment by seberg** [comment]
It is a bit tricky, the easiest way is this, IMO: `np.empty(5, dtype=np.uint8)[1:].view(np.int32)`.

**Comment by robbmcleod** [comment]
Yes this fails:

    x = np.empty(5, dtype=np.uint8)[1:].view(np.int32)
    print(f"pre-eval, x = {x} at {x.__array_interface__['data'][0]}")
    ne.evaluate('3', out=x)
    print(f"post-eval, x = {x} at {x.__array_interface__['data'][0]}")

outputs:

    pre-eval, x = [217964907] at 2254816281185
    n_inputs = 0
    operands[0] = 0000020CFDE699F0, retsig = i
    Case with NPY_ARRAY_UPDATEIFCOPY
    post-eval, x = [217964907] at 2254816281185

So it is no longer setting x to 3. Any suggestions on possible fixes? I am pretty familiar with the NumPy C-API but not the flag issues here.

**Comment by seberg** [comment]
Basically, you can just use `NPY_ARRAY_WRITEBACKIFCOPY` and it should work for CPython.  To be clean and on PyPy, you have to call `PyArray_ResolveWritebackIfCopy` before returning to the user.  See here:

https://numpy.org/devdocs/release/1.14.0-notes.html#pypy-compatible-alternative-to-updateifcopy-arrays

**Comment by mattip** [comment]
If you run python with -Walways or -Werror, and run the failing test with an older NumPy, do you see the DeprecationWarning?

**Comment by robbmcleod** [comment]
@mattip No I don't get any warning from NumPy with `-Wall` with `NPY_ARRAY_UPDATEIFCOPY`. I do get a warning if  I use `NPY_ARRAY_WRITEBACKIFCOPY`. 

Thinking about it some more, because this code branch only executes in a single case of a single constant element, it may be more efficient to simply set the value in the branch and then use a `goto` to bypass all the NumExpr machinery. Although the path of least resistance would be to just use `NPY_ARRAY_UPDATEIFCOPY` and see if anyone raises an issue with me not handling the array destruction properly.

**Comment by rgommers** [comment]
This is broken now that numpy 1.23.0 has been released, so may warrant a new `numexpr` release?

**Comment by seberg** [comment]
How important is this?  We can milestone it for the .1 release.

**Comment by seberg** [comment]
Hmmm, I thought there may be a way to work around in NumPy for the 1.23.x series, the removal was not a massive change anyway, even putting it back may be possible.
(I have not looked into that in detail, but if that may be interesting we need a milestoned issue on NumPy, or I will definitely forget.)

**Comment by robbmcleod** [comment]
Mkay, sorry, I have been a bit busy lately. I'll look at it today.

**Comment by robbmcleod** [comment]
What are the consequences of ignoring this warning: 

` RuntimeWarning: WRITEBACKIFCOPY detected in array_dealloc.  Required call to PyArray_ResolveWritebackIfCopy or PyArray_DiscardWritebackIfCopy is missing.`

Does NumPy leak a little memory here? We can only have this issue for singleton arrays.

**Comment by seberg** [comment]
NumPy may be leaking a reference here, if that is what matters to you, but no its not a memory leak.

The problem is that when running in PyPy, the mechanism may fail and users will see uninitialized memory instead of the result.

**Comment by robbmcleod** [comment]
If I use no flags, i.e. don't enforce use of an aligned array, it seems to work fine:

    a = (PyArrayObject *)PyArray_FromArray(operands[0], dtypes[0], 0);

Not sure what, if anything is wrong with that. I think whomever originally implemented this branch thought the array needed to be aligned, but it really doesn't. Given that it's a singleton, there's no performance issues at stake.

**Comment by seberg** [comment]
@robbmcleod but on certain platforms the alignment may be necessary, unless  you are very careful in `run_interpreter_const` (which might be the case)?

I am not sure why you cannot just call `PyArray_ResolveWritebackIfCopy` after the call to `run_interpreter_const `?  I think you can call that even if it is not used at all, but a switch would also work.

**Comment by robbmcleod** [comment]
@seberg if that is practical then I was confused about when that function needed to be called. I thought it needed to be called on array destruction.  

I tried implementing it but it failed for the Python 3.6 build. 

https://github.com/pydata/numexpr/runs/7032987348

Looks like  `NPY_ARRAY_WRITEBACKIFCOPY` was introduced in 1.14.0? And `oldest_supported_numpy` is providing `1.13.3`... So either drop Python 3.6 support or write some `#define`s in to the code. I guess dropping 3.6 is fine.

**Comment by seberg** [comment]
Ouch!  I guess you could also add a manual `>=1.14` version requirement for python3.6 if you don't care much?

Btw, I think there is still one oddity with the code: The return value is the new array, when it should not be:
```
import numpy as np
import numexpr as ne

x = np.empty(5, dtype=np.uint8)[1:].view(np.int32)
r = ne.evaluate('3', out=x)
r is x  # False
```
not sure you care about that, but the res should probably be defined in the earlier branch if you want to avoid that.  (It does not matter as much with `WriteBackIfCopy` though).

Ar you OK with that version restriction or not?

**Comment by robbmcleod** [comment]
@seberg and @rgommers let me know tomorrow if you see any issues with the code section from line 1250 onward in `intepreter.cpp`, as I'm ready to release here. There were a few commits so I provide the latest code here:

```
    if (n_inputs == 0) {
        char retsig = get_return_sig(self->program);

        /* Allocate the output */
        if (operands[0] == NULL) {
            npy_intp dim = 1;
            operands[0] = (PyArrayObject *)PyArray_SimpleNew(0, &dim,
                                        typecode_from_char(retsig));
            if (operands[0] == NULL) {
                goto fail;
            }
        }
        else { // Use the provided output array
            if (PyArray_SIZE(operands[0]) != 1) {
                PyErr_SetString(PyExc_ValueError,
                        "output for a constant expression must have size 1");
                goto fail;
            }
            else if (!PyArray_ISWRITEABLE(operands[0])) {
                PyErr_SetString(PyExc_ValueError,
                        "output is not writeable");
                goto fail;
            }
            Py_INCREF(dtypes[0]);

            // NumPy folks suggested using WRITEBACKIFCOPY to resolve issue #397
            singleton = (PyArrayObject *)PyArray_FromArray(operands[0], dtypes[0],
                                        NPY_ARRAY_ALIGNED|NPY_ARRAY_WRITEBACKIFCOPY);

            if (singleton == NULL) {
                goto fail;
            }
            Py_DECREF(operands[0]);
            operands[0] = singleton;
        }

        r = run_interpreter_const(self, PyArray_BYTES(operands[0]), &pc_error);

        if ((n_inputs == 0) && (operands[0] != NULL)) {
            // Write-back our copy to the passed in output array if we had to make a copy.
            int retval = PyArray_ResolveWritebackIfCopy(singleton);
            if (retval < 0) {
                // 1 means it copied the value, 0 means no copy, only -1 is an error.
                PyErr_Format(PyExc_ValueError, "Writeback to singleton failed with error code: %d", retval);
                goto fail;
            }
        }
        ret = (PyObject *)operands[0];
        Py_INCREF(ret);
        goto cleanup_and_exit;
    }

```

**Comment by seberg** [comment]
Look good, I assume the tests pass? It seems like the `operands[0] != NULL` may be there to check whether an input was passed and the `n_inputs` one is also always true?
Both seem no-ops if they are there, but it should not matter much.

**Comment by robbmcleod** [comment]
It's explicit in case any changes are made in the future to either NumPy or NumExpr that `PyArray_ResolveWritebackIfCopy` is only required on that particular branch of code.

**Comment by seberg** [comment]
I like the explicit check, but operands always gets set here, so it is always true?

**Comment by robbmcleod** [comment]
Alright I've released 2.8.3. I skipped 2.8.2 because I once again didn't have the right description metadata for RST. Not sure how that happened, probably I didn't pull remote changes at some point. In any case I think this is resolved, if people have trouble with the writeback please re-open.
